File: src/com/sterndu/multicore/MultiCore.java
package com.sterndu.multicore;

import java.lang.Thread.State;
import java.util.*;
import java.util.Map.Entry;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.stream.Stream;
import com.sterndu.util.interfaces.ThrowingConsumer;

public class MultiCore {

	// TODO max threads

	public static abstract class TaskHandler {
		protected double prioMult;
		protected double lastAverage;
		protected final List<Long> times;

		protected TaskHandler() {
			prioMult=.1d;
			lastAverage=.0d;
			times = new ArrayList<>();
		}

		protected TaskHandler(double prioMult) {
			this.prioMult=prioMult;
			lastAverage=.0d;
			times = new ArrayList<>();
		}

		protected void addTime(long time) {
			synchronized (times) {
				times.add(time);
				if (times.size() > 30)
					times.remove(0);
			}
		}

		protected abstract ThrowingConsumer<TaskHandler> getTask();

		protected abstract boolean hasTask();

		public double getAverageTime() {
			synchronized (times) {
				return lastAverage = times.parallelStream().mapToLong(l -> l).average().getAsDouble();
			}
		}

		public double getLastAverageTime() {
			return lastAverage;
		}

		public double getPrioMult() {
			return prioMult;
		}

		public List<Long> getTimes(){
			return new ArrayList<>(times);
		}
	}

	private static MultiCore multiCore;
	static {
		multiCore = new MultiCore();
	}

	private final Object simThreadsLock = new Object();
	private int simultaneousThreads = 0;

	private final AtomicBoolean ab;
	private int count;

	private final List<TaskHandler> taskHandler;

	private final Thread[] threads;
	private final Runnable r;

	private MultiCore() {
		r = () -> {
			while (true) {
				Entry<TaskHandler, ThrowingConsumer<TaskHandler>> data = multiCore.getTask();
				if (data == null)
					break;
				ThrowingConsumer<TaskHandler> data2 = data.getValue();
				if (data2 != null) {
					long st = System.currentTimeMillis();
					try {
						data2.accept(data.getKey());
						long et = System.currentTimeMillis();
						et -= st;
						data.getKey().addTime(et);
						data.getKey().getAverageTime();
					} catch (Exception e) {
						e.printStackTrace();
					}
				} else break;
				try {
					Thread.sleep(1);
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		};
		ab = new AtomicBoolean(false);
		taskHandler = new ArrayList<>();
		threads=new Thread[Runtime.getRuntime().availableProcessors()];
		for (int i = 0; i < threads.length; i++) threads[i] = new Thread(r, "MultiCore-Worker=" + i);
		Runtime.getRuntime().addShutdownHook(new Thread(() -> {
			close();
		}));
	}

	private static int checkIfMoreThreadsAreRequiredAndStartSomeIfNeeded() {
		int sum = getAmountOfAvailableTasks();
		if (sum > 0) synchronized (multiCore.simThreadsLock) {
			setSimultaneousThreads(getSimultaneousThreads(), sum);
		}
		else if (getActiveThreadsCount() == 0) setSimultaneousThreads(getSimultaneousThreads(), 1);
		return sum;
	}

	private static int getActiveThreadsCount() {
		return (int) Stream.of(MultiCore.multiCore.threads).filter(Thread::isAlive).count();
	}

	private static void reSort() {
		synchronized (multiCore.taskHandler) {
			Collections.sort(multiCore.taskHandler, (d1, d2) -> Double.compare(d2.getLastAverageTime() * d2.prioMult,
					d1.getLastAverageTime() * d1.prioMult));
		}
	}

	public static void addTaskHandler(TaskHandler taskHandler) {
		synchronized (multiCore.taskHandler) {
			multiCore.taskHandler.add(taskHandler);
			reSort();
		}
		checkIfMoreThreadsAreRequiredAndStartSomeIfNeeded();

	}

	public static void close() {
		multiCore.ab.set(true);
	}

	public static int getAmountOfAvailableTasks() {
		synchronized (multiCore.taskHandler) {
			return multiCore.taskHandler.parallelStream().mapToInt(t -> t.hasTask() ? 1 : 0).sum();
		}
	}

	public static int getSimultaneousThreads() {
		return multiCore.simultaneousThreads;
	}

	public static boolean removeTaskHandler(TaskHandler taskHandler) {
		synchronized (multiCore.taskHandler) {
			boolean b = multiCore.taskHandler.remove(taskHandler);
			reSort();
			return b;
		}

	}

	public static void setSimultaneousThreads(int amount, int... data) {
		synchronized (multiCore.simThreadsLock) {
			multiCore.simultaneousThreads = Math.min(multiCore.threads.length, amount);
			for (int i = 0; i < multiCore.threads.length; i++)
				if (multiCore.threads[i].getState().equals(State.TERMINATED))
					multiCore.threads[i] = new Thread(multiCore.r, "MultiCore-Worker=" + i);
			int temp = Math.max(multiCore.simultaneousThreads,
					data != null && data.length > 0 ? data[0] : multiCore.simultaneousThreads);
			if (MultiCore.getActiveThreadsCount() < temp) {
				int activate = temp - MultiCore.getActiveThreadsCount();
				for (Thread th: multiCore.threads) {
					if (!th.isAlive()) {
						th.start();
						activate--;
					}
					if (activate == 0)
						break;
				}
			}
		}
	}

	private synchronized Entry<TaskHandler, ThrowingConsumer<TaskHandler>> getTask() {

		if (checkIfMoreThreadsAreRequiredAndStartSomeIfNeeded() > 0) {
			if (count == taskHandler.size() - 1) count = 0;
			else count++;
			TaskHandler handler = taskHandler.get(count);
			if (handler.hasTask()) return Map.entry(handler, handler.getTask());
		} else if (ab.get() | MultiCore.getActiveThreadsCount() > 1) return null;
		try {
			Thread.sleep(2);
		} catch (InterruptedException e) {
			e.printStackTrace();
			return null;
		}
		return Map.entry(new TaskHandler() {

			@Override
			protected ThrowingConsumer<TaskHandler> getTask() { return null; }

			@Override
			protected boolean hasTask() {
				return false;
			}
		}, th -> {});
	}

}

File: src/com/sterndu/multicore/MultiCore.kt
@file:JvmName("MultiCore")
package com.sterndu.multicore

import java.util.concurrent.*
import java.util.logging.Level

object MultiCore {

	val logger = LoggingUtil.getLogger("MultiCore")

	private val ses = Executors.newScheduledThreadPool(
		Runtime.getRuntime().availableProcessors(),
		Thread.ofVirtual().factory()
	) as ScheduledThreadPoolExecutor

	private val taskHandlers: MutableList<TaskHandler> = CopyOnWriteArrayList()

	private val scheduledTasks: MutableMap<Any, ScheduledFuture<*>> = HashMap()

	private val kernel: (TaskHandler, Runnable) -> Unit = { key, data ->
		val st = System.currentTimeMillis()
		try {
			data.run()
			var et = System.currentTimeMillis()
			et -= st
			key.addTime(et)
		} catch (e: Exception) {
			e.printStackTrace()
		}
	}

	init {
		ses.allowCoreThreadTimeOut(true)
		ses.scheduleWithFixedDelay({
			// DEBUG Updater.logger.info("Doing House keeping")
			try {
				for (taskHandler in taskHandlers) {
					when (taskHandler) {
						is RepeatingTaskHandler -> {
							taskHandler.taskInformationMap
								.map(Map.Entry<Any, RepeatingTaskHandler.Information>::value)
								.filter { it.runnable !in scheduledTasks }
								.forEach { information ->
									val future = ses.scheduleWithFixedDelay(
										{ kernel(taskHandler) { taskHandler.r(information) } },
										0,
										information.millis.coerceAtLeast(1),
										TimeUnit.MILLISECONDS
									)

									scheduledTasks[information.runnable] = future
								}
							cleanupNonExistentTasks(taskHandler)
						}

						else -> {
							while (taskHandler.hasTask()) {
								taskHandler.internalGetTask()?.let { task ->
									scheduledTasks[task] = ses.schedule({ kernel(taskHandler, task) }, 0, TimeUnit.MILLISECONDS)
								}
							}
						}
					}
				}
				if (scheduledTasks.entries.removeIf { (_, future) -> future.isDone || future.isCancelled } && "true" == System.getProperty("debug")) {
					logger.info("Removed a task")
				}

			} catch (e: Exception) {
				logger.log(Level.WARNING, "MultiCore ${e.javaClass.simpleName} ${e.message} ${e.cause}", e)
			}
		}, 0, 1, TimeUnit.MILLISECONDS)
		
		Runtime.getRuntime().addShutdownHook(Thread { stop() })
	}

	private fun cleanupNonExistentTasks(taskHandler: Updater) {
		scheduledTasks
			.map(Map.Entry<Any, ScheduledFuture<*>>::key)
			.filterIsInstance<Runnable>()
			.filterNot {
				taskHandler.taskInformationMap
					.map(Map.Entry<Any, RepeatingTaskHandler.Information>::value)
					.map(RepeatingTaskHandler.Information::runnable)
					.any(it::equals)
			}
			.forEach {
				// DEBUG Updater.logger.info("Cleaned a task")
				scheduledTasks[it]!!.cancel(false)
				scheduledTasks.remove(it)
			}
	}

	private fun checkIfMoreThreadsAreRequiredAndStartSomeIfNeeded(): Int {
		return amountOfAvailableTasks
	}

	private val activeThreadsCount: Int
		get() = ses.activeCount

	fun addTaskHandler(taskHandler: TaskHandler) {
		this.taskHandlers.add(taskHandler)
		checkIfMoreThreadsAreRequiredAndStartSomeIfNeeded()
	}

	fun stop() {
		ses.shutdown()
	}

	val amountOfAvailableTasks: Int
		get() {
			return this.taskHandlers.parallelStream()
				.mapToInt { t: TaskHandler -> if (t.hasTask()) 1 else 0 }
				.sum()
		}

	fun getSimultaneousThreads(): Int {
		return ses.maximumPoolSize
	}

	fun getActiveThreads() = ses.activeCount

	fun removeTaskHandler(taskHandler: TaskHandler): Boolean {
		return taskHandlers.remove(taskHandler)
	}

	@Synchronized
	fun setSimultaneousThreads(amount: Int) {
		ses.corePoolSize = amount
	}
}
